provider/oracle/environ.go

```diff
@@ -0,0 +1,744 @@
+	}
+
+	tools, err := args.Tools.Match(tools.Filter{Arch: spec.Image.Arch})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Tracef("Tools: %v", tools)
+	if err = args.InstanceConfig.SetTools(tools); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if err = instancecfg.FinishInstanceConfig(
+		args.InstanceConfig,
+		o.Config(),
+	); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// new cloud config template based on the series
+	cloudcfg, err := cloudinit.New(args.InstanceConfig.Series)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot create cloudinit template")
+	}
+
+	// compose userdata with the cloud config template
+	userData, err := providerinit.ComposeUserData(
+		args.InstanceConfig,
+		cloudcfg,
+		OracleRenderer{},
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot make user data")
+	}
+
+	logger.Debugf("oracle user data: %d bytes", len(userData))
+
+	attributes := map[string]interface{}{
+		"userdata": string(userData),
+	}
+
+	hostname, err := o.namespace.Hostname(args.InstanceConfig.MachineId)
+	machineName := o.client.ComposeName(hostname)
+	imageName := o.client.ComposeName(imagelist)
+
+	tags := make([]string, 0, len(args.InstanceConfig.Tags)+1)
+	for k, v := range args.InstanceConfig.Tags {
+		if k == "" || v == "" {
+			continue
+		}
+		t := tagValue{k, v}
+		tags = append(tags, t.String())
+	}
+	tags = append(tags, machineName)
+
+	var apiPort int
+	if args.InstanceConfig.Controller != nil {
+		apiPort = args.InstanceConfig.Controller.Config.APIPort()
+	} else {
+		// All ports are the same so pick the first.
+		apiPort = args.InstanceConfig.APIInfo.Ports()[0]
+	}
+
+	// create a new seclists
+	secLists, err := o.firewall.CreateMachineSecLists(
+		args.InstanceConfig.MachineId, apiPort)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// create a  new netowrking card used for making the instance
+	// have a public address ip
+	networking := map[string]oci.Networker{
+		"eth0": oci.SharedNetwork{
+			Seclists: secLists,
+		},
+	}
+
+	// create the instance based on the instance params
+	instance, err := o.createInstance(o.client, oci.InstanceParams{
+		Instances: []oci.Instances{
+			{
+				Shape:       spec.InstanceType.Name,
+				Imagelist:   imageName,
+				Name:        machineName,
+				Label:       args.InstanceConfig.MachineAgentServiceName,
+				Hostname:    args.InstanceConfig.MachineAgentServiceName,
+				Tags:        tags,
+				Attributes:  attributes,
+				Reverse_dns: false,
+				Networking:  networking,
+			},
+		},
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	instance.arch = &spec.Image.Arch
+	instance.instType = &spec.InstanceType
+
+	machineId := instance.machine.Name
+	timeout := 10 * time.Minute
+	if err := instance.waitForMachineStatus(ociCommon.StateRunning, timeout); err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Infof("started instance %q", machineId)
+	//TODO: add config option for public IP allocation
```

Comment:
Having an option is fine. I think we should default to giving a public IP, though, since that's what all other public clouds are doing at the moment. We may want to revisit that policy later. I agree that it's wasteful to give every machine one by default. IMO, what we *should* be doing is giving them a public IP only when the machine contains a unit whose application is exposed. One issue with that is that you won't be able to SSH in without specifying "--proxy", which is kinda poor UX.